### PR TITLE
ovmf: expose EFI prefixes and refactor qemu-vm with it

### PR DESCRIPTION
**Pull request by RaitoBezarius**
###### Description of changes

In summary:
- EDK2 `mkDerivation` supports functional-style `mkDerivation` ;
- OVMF exposes EFI prefixes in `passthru` as `firmware` and `variables`
- `virtualisation/qemu-vm` uses it to pass them in the machinery, consolidating it ;
- adding myself as a maintainer of OVMF as I know well the ecosystem and am interested into maintaining it

**How to test this?** : `nix-build nixos/tests/systemd-boot.nix -A basic`, it if works, then QEMU's EFI still works fine. :)

It is pretty common to require the EFI firmware and variables of OVMF, this exposes it in `passthru` as suggested by @roberth on Matrix.
`qemu-vm` can be refactored with this, showing an example of usage.
This makes it easier to consolidate and add new architectures that OVMF supports, e.g. RISC-V, Raspberry Pi, etc.
I tested `systemd-boot.basic` test with this, which relies on EFI, so it should have tested this part of the code.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 1]
```diff
@@ -98,12 +98,9 @@ let
...
 
-  efiPrefix =
-    if pkgs.stdenv.hostPlatform.isx86 then "${pkgs.OVMF.fd}/FV/OVMF"
-    else if pkgs.stdenv.isAarch64 then "${pkgs.OVMF.fd}/FV/AAVMF"
-    else throw "No EFI firmware available for platform";
-  efiFirmware = "${efiPrefix}_CODE.fd";
-  efiVarsDefault = "${efiPrefix}_VARS.fd";
+  efiPrefix = pkgs.OVMF.getEfiPrefix pkgs.stdenv.hostPlatform;
+  efiFirmware = efiPrefix.firmware;
+  efiVarsDefault = efiPrefix.variables;
```
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/qemu-vm.nix:103]: We might as well inline efiFirmware and efiVarsDefault to reduce the layers of redirection.

File: pkgs/applications/virtualization/OVMF/default.nix [hunk 2]
```diff
@@ -62,10 +68,21 @@ edk2.mkDerivation projectDscPath {
 
   dontPatchELF = true;
 
+  passthru.getEfiPrefix = platform:
```
Comment (roberth) [inline pkgs/applications/virtualization/OVMF/default.nix:71]: You could do

```suggestion
  passthru.getEfiPrefix = { platform ? pkgs.stdenv.hostPlatform }:
```

but maybe that's over-engineered, if you only need

```nix
  passthru.efiPrefix = let platform = pkgs.stdenv.hostPlatform; in
```

I'm inclined to assume that if the user wants to set `platform`, they've already set `hostPlatform` for cross compilation.

Then you can also simplify to:

```nix
  passthru = { firmware = ...; variables = ...; };
```

I am completely unfamiliar with this package though, so I don't really know.

You'd be free to add `getEfiPrefix` later if there's a real need.

File: pkgs/development/compilers/edk2/default.nix [hunk 3]
```diff
@@ -87,7 +87,11 @@ edk2 = buildStdenv.mkDerivation {
   };
 
   passthru = {
-    mkDerivation = projectDscPath: attrs: buildStdenv.mkDerivation ({
+    mkDerivation = projectDscPath: attrsOrFun: buildStdenv.mkDerivation (finalAttrs:
+    let
+      attrs = if lib.isFunction attrsOrFun then (attrsOrFun finalAttrs) else attrsOrFun;
```
Comment (roberth) [inline pkgs/development/compilers/edk2/default.nix:92]:
```suggestion
      attrs = lib.toFunction attrsOrFun finalAttrs;
```

File: pkgs/applications/virtualization/OVMF/default.nix (hunk 2 above)
Comment (RaitoBezarius) [reply pkgs/applications/virtualization/OVMF/default.nix:71]: Right, I think it's definitely better to go with your suggestion. We can always do more complicated stuff later.

File: nixos/modules/virtualisation/qemu-vm.nix (hunk 1 above)
Comment (RaitoBezarius) [reply nixos/modules/virtualisation/qemu-vm.nix:103]: You're right

**Review by danielfullmer — APPROVED** [review]
LGTM, Thanks!

File: pkgs/applications/virtualization/OVMF/default.nix (hunk 2 above)
Comment (RaitoBezarius) [reply pkgs/applications/virtualization/OVMF/default.nix]: Done.

File: nixos/modules/virtualisation/qemu-vm.nix (hunk 1 above)
Comment (RaitoBezarius) [reply nixos/modules/virtualisation/qemu-vm.nix]: Done.

File: pkgs/applications/virtualization/OVMF/default.nix [hunk 4]
```diff
@@ -62,10 +68,21 @@ edk2.mkDerivation projectDscPath {
 
   dontPatchELF = true;
 
+  passthru =
+  let
+    cpuName = stdenv.hostPlatform.parsed.cpu.name;
+    suffix = suffixes."${cpuName}" or (throw "`${cpuName}` is not supported in this OVMF derivation!");
```
Comment (roberth) [inline pkgs/applications/virtualization/OVMF/default.nix:74]:
```suggestion
    suffix = suffixes."${cpuName}" or (throw "Host cpu name `${cpuName}` is not supported in this OVMF derivation!");
```

**Comment by roberth** [comment]
Could you add some tests to `passthru.tests` in the `edk2` and `ovmf` packages?

**Comment by RaitoBezarius** [comment]
> Could you add some tests to `passthru.tests` in the `edk2` and `ovmf` packages?
> 
> I assume we already have tests for this, but I don't know.

We do not have direct tests for this, except that using EFI in NixOS tests are using ovmf (and then edk2) indirectly, would you like me to add the indirect tests?

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 5]
```diff
@@ -98,12 +98,10 @@ let
   addDeviceNames =
     imap1 (idx: drive: drive // { device = driveDeviceName idx; });
 
-  efiPrefix =
-    if pkgs.stdenv.hostPlatform.isx86 then "${pkgs.OVMF.fd}/FV/OVMF"
-    else if pkgs.stdenv.isAarch64 then "${pkgs.OVMF.fd}/FV/AAVMF"
-    else throw "No EFI firmware available for platform";
-  efiFirmware = "${efiPrefix}_CODE.fd";
-  efiVarsDefault = "${efiPrefix}_VARS.fd";
+  # TODO: support alternatives to OVMF implementations
```
[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply nixos/modules/virtualisation/qemu-vm.nix]: Yes, I do plan to tackle it for RISC-V64. I created the new two options.

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 6]
```diff
@@ -98,12 +98,10 @@ let
...
+  efiPrefix = {
```
[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply nixos/modules/virtualisation/qemu-vm.nix]: (no need anymore as moved to options)

File: pkgs/applications/virtualization/OVMF/default.nix (hunk 4 above)
Comment (RaitoBezarius) [reply pkgs/applications/virtualization/OVMF/default.nix]: Done.

**Comment by roberth** [comment]
OfBorg doesn't seem to pick up on the lowercase commit message prefix.

@ofborg build ovmf

**Comment by RaitoBezarius** [comment]
@roberth unclear if https://logs.nix.ci/?key=nixos/nixpkgs.187887&attempt_id=a3efc083-4804-4b47-b326-aef7d9d3c51b is a failure or not to me.

**Comment by roberth** [comment]
> @roberth unclear if https://logs.nix.ci/?key=nixos/nixpkgs.187887&attempt_id=a3efc083-4804-4b47-b326-aef7d9d3c51b is a failure or not to me.

It's a clean refactor producing the exact same derivation, so the problem isn't caused by this PR.

**Review by roberth — APPROVED** [review]

**Pull request merged**